lib/THC/THCTensorMasked.cu

```diff
@@ -44,11 +42,91 @@ void THCudaTensor_maskedFill(THCState* state,
   THCudaCheck(cudaGetLastError());
 }
 
+struct TensorMaskedCopyOp {
+  TensorMaskedCopyOp(float* s, float* bm, float* ps)
+      : src(s),
+        baseMask(bm),
+        maskPrefixSum(ps) {
+  }
+
+  __device__ __forceinline__ void operator()(float* out, float* mask) {
+    // Really mask should be `0` or `1` but we can't propagate errors here.
+    if (*mask != 0.0f) {
+      // We've already checked that this offset is <= 2^24, so this is ok.
+      int srcOffset = (int) (mask - baseMask);
+      *out = src[(int) maskPrefixSum[srcOffset]];
+    }
```

Comment:
Can we not use `THCudaTensor_pointwiseApply3` here and pass in `maskPrefixSum` as an argument? Then we wouldn't need to calculate the offset.